packages/bns/src/index.ts

```diff
@@ -0,0 +1,764 @@
+ * @param  {String} salt - salt used to generate the preorder namespace hash
+ * @param  {BigNum} stxToBurn - amount of STX to burn for the registration
+ * @param  {String} publicKey - the private key to sign the transaction
+ * @param  {StacksNetwork} network - the Stacks blockchain network to register on
+ */
+export interface PreorderNamespaceOptions {
+  namespace: string;
+  salt: string;
+  stxToBurn: BN;
+  publicKey: string;
+  network?: StacksNetwork;
+}
+
+/**
+ * Generates a namespace preorder transaction.
+ * First step in registering a namespace. This transaction does not reveal the namespace that is
+ * about to be registered. And it sets the amount of STX to be burned for the registration.
+ *
+ * Resolves to the generated StacksTransaction
+ *
+ * @param  {PreorderNamespaceOptions} options - an options object for the preorder
+ *
+ * @return {Promise<StacksTransaction>}
+ */
+export async function buildPreorderNamespaceTX({
+  namespace,
+  salt,
+  stxToBurn,
+  publicKey,
+  network,
+}: PreorderNamespaceOptions): Promise<StacksTransaction> {
+  const bnsFunctionName = 'namespace-preorder';
+  const saltedNamespaceBuffer = Buffer.from(`0x${namespace}${salt}`);
+  const hashedSaltedNamespace = hash160(saltedNamespaceBuffer);
+  const txNetwork = network || new StacksMainnet();
+
+  return makeBNSContractCall({
+    functionName: bnsFunctionName,
+    functionArgs: [bufferCV(hashedSaltedNamespace), uintCVFromBN(stxToBurn)],
+    publicKey,
+    network: txNetwork,
+  });
+}
+
+/**
+ * Reveal namespace options
+ *
+ * @param  {String} namespace - the namespace to reveal
+ * @param  {String} salt - salt used to generate the preorder namespace hash
+ * @param  {PriceFunction} priceFunction - an object containing the price function for the namespace
+ * @param  {BigNum} lifeTime - the number of blocks name registrations are valid for in the namespace
+ * @param  {String} namespaceImportAddress - the STX address used for name imports
+ * @param  {String} publicKey - the private key to sign the transaction
+ * @param  {StacksNetwork} network - the Stacks blockchain network to register on
+ */
+export interface RevealNamespaceOptions {
+  namespace: string;
+  salt: string;
+  priceFunction: PriceFunction;
+  lifetime: BN;
+  namespaceImportAddress: string;
+  publicKey: string;
+  network?: StacksNetwork;
+}
+
+/**
+ * Generates a namespace reveal transaction.
+ * Second step in registering a namespace.
+ *
+ * Resolves to the generated StacksTransaction
+ *
+ * @param  {RevealNamespaceOptions} options - an options object for the reveal
+ *
+ * @return {Promise<StacksTransaction>}
+ */
+export async function buildRevealNamespaceTX({
+  namespace,
+  salt,
+  priceFunction,
+  lifetime,
+  namespaceImportAddress,
+  publicKey,
+  network,
+}: RevealNamespaceOptions): Promise<StacksTransaction> {
+  const bnsFunctionName = 'namespace-reveal';
+  const txNetwork = network || new StacksMainnet();
```

Comment:
Is setting defaults like this a safe pattern? IMO it'd better to require this to be set explicitly.